### PR TITLE
feat: add completion to ':lua'

### PR DESCRIPTION
**Pull request by tjdevries**
@bfredl THOUGHTS??? YOU DOUBTED ME?!?!?

https://clips.twitch.tv/EnchantingImpartialCheesecakePunchTrees

^^^ SELF DOCUMENTING

### CODE REVIEW

**Comment by sunjon** [comment]
Great work!
A couple of bits of feedback:
- It it intended that the Pum message is showing more than the current word you're completing?
- (I just realised that the second thing I was going to ask for is because I've become spoiled by builtin LSP)
  Will it support `foo.x<tab>`, `foo:x<tab>`? ;]

**Comment by bfredl** [comment]
@tjdevries looks nice. :) I don't think "my" completer (mostly stolen from luatorch) has the prefix issue, I could compare our implementations and see how to get rid of it.

File: test/functional/lua/command_line_completion_spec.lua [hunk 1]
```diff
@@ -0,0 +1,166 @@
+local helpers = require('test.functional.helpers')(after_each)
+
+local clear = helpers.clear
+local eq = helpers.eq
+local funcs = helpers.funcs
+
+local get_completions = function(input, env)
+  return funcs.luaeval("vim._expand_pat(_A.input, _A.env)", { input = '^' .. input, env = env })
```
Comment (bfredl) [inline test/functional/lua/command_line_completion_spec.lua:8]: use `exec_lua("return vim._expand_pat(...)", '^' .. input,  env)` like a PRO.

File: test/functional/lua/command_line_completion_spec.lua [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    it('should allows back-to-back string ref with spaces before and after', function()
+      eq({{"vim", "nested", "api"}, 25}, get_compl_parts('vim[ "nested"  ]["api"].nvim_buf'))
+    end)
+
+    it('should allow VAR style loolup', function()
+      eq({{"vim", {"NESTED"}, "api"}, 20}, get_compl_parts('vim[NESTED]["api"].nvim_buf'))
+    end)
+  end)
+end)
```
Comment (bfredl) [inline test/functional/lua/command_line_completion_spec.lua:166]: we also want an integration test for `ex_getln.c` integration. Could look into `viml/completion_spec.lua` and how it does screen tests.

File: src/nvim/lua/executor.c [hunk 3]
```diff
@@ -1264,6 +1264,59 @@ static void nlua_add_treesitter(lua_State *const lstate) FUNC_ATTR_NONNULL_ALL
...
+    GA_APPEND(char_u *, &result_array, (char_u *)v.data.string.data);
+  }
+
+  *results = result_array.ga_data;
+  *num_results = result_array.ga_len;
+
+  return OK;
+
+cleanup:
+  ga_clear(&result_array);
```
Comment (bfredl) [inline src/nvim/lua/executor.c:1315]: do you even `api_free_array(completions)` ?

Comment (tjdevries) [reply src/nvim/lua/executor.c:1315]: Do I need it for both cases (i.e. with an error and w/out?)

Comment (bfredl) [reply src/nvim/lua/executor.c:1315]: Yes.

Comment (tjdevries) [reply src/nvim/lua/executor.c:1315]: :+1: OK, sounds good. 

the `ga_*` stuff gets cleared automatically by the parent completion stuff, right?

Comment (bfredl) [reply src/nvim/lua/executor.c:1315]: I think so, but a complete integration test is gonna confirm that on CI :)

**Comment by tjdevries** [comment]
Idk how to fix the completer thing -- I prepend the other text, but only because otherwise the command line overwrites everything on the line with ONLY the suggestion.

So I must be missing something about the completion there.

**Comment by lithammer** [comment]
I haven't tested this myself. But judging by the screenshots and video clips, it looks like it's not using the same mechanism as the regular tab completion for vim script. Meaning no wildmenu support etc. It just looks like the regular ins-completion-menu which doesn't seem right..?

**Comment by gravndal** [comment]
@lithammer You might want to have a look at `:h 'wildoptions'` or https://github.com/neovim/neovim/pull/9607 and https://github.com/neovim/neovim/pull/10384

**Comment by bfredl** [comment]
@tjdevries hmm it is a bit tricky but it should be possible to change (i know because i did it for path completion in `wop=pum` )

**Comment by tjdevries** [comment]
You have any idea where the code is? I'll take a look later.

**Comment by tjdevries** [comment]
No longer needed, closed by #13719

**Pull request closed without merging**